File: python/largest-series-product/series.py
from functools import reduce
from operator import mul

__author__ = 'Cedric Zhuang'


def largest_product(sequence, length):
    groups = slices(sequence, length)
    return max((reduce(mul, group, 1) for group in groups))


def slices(sequence, length):
    slen = len(sequence)
    if length > slen or slen > length == 0:
        raise ValueError()
    ints = list(map(int, list(sequence)))
    return [ints[index:index + length]
            for index in range(len(ints) - length + 1)]

File: scala/largest-series-product/src/main/scala/Series.scala
/**
  *
  * @author Cedric Zhuang
  *
  */
object Series {
  def largestProduct(i: Int, s: String): Option[Int] =
    if (i == 0) Some(1)
    else if (s.isEmpty || i > s.length) None
    else Some(calculate(i, s))

  private def calculate(i: Int, s: String): Int =
    s.map(_ - '0').sliding(i).map(_.product).max
}
